### PR TITLE
Compact enum store dictionary when needed.

### PR DESCRIPTION
**Pull request by toregge**
@baldersheim : please review

### CODE REVIEW

**Review by baldersheim — APPROVED** [review]
This looks good to me, but I think that @geirst should take a closer look. He knows this area better than me.

File: vespalib/src/vespa/vespalib/datastore/sharded_hash_map.cpp [hunk 1]
```diff
@@ -194,4 +194,30 @@ ShardedHashMap::normalize_values(std::function<EntryRef(EntryRef)> normalize)
...
 }
 
+bool
+ShardedHashMap::has_held_buffers() const
+{
+    return _gen_holder.getHeldBytes() != 0;
+}
+
+void
+ShardedHashMap::compact_worst()
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/datastore/sharded_hash_map.cpp:204]: Consider adding explicit unit test for this in sharded_hash_map_test.cpp.

**Review by geirst — APPROVED** [review]
👍

**Review by geirst — APPROVED** [review]
👍

File: searchlib/src/vespa/searchlib/attribute/enumstore.hpp [hunk 2]
```diff
@@ -242,6 +245,24 @@ EnumStoreT<EntryT>::compact_worst_values(bool compact_memory, bool compact_addre
...
 }
 
+template <typename EntryT>
+bool
+EnumStoreT<EntryT>::consider_compact_dictionary(const CompactionStrategy& compaction_strategy)
+{
+    if (_dict->has_held_buffers()) {
+        return false;
+    }
+    if (compaction_strategy.should_compact_memory(_cached_dictionary_btree_usage.usedBytes(), _cached_dictionary_btree_usage.usedBytes())) {
```
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/attribute/enumstore.hpp:255]: Ops, I think we missed that all are not dead by default.

File: searchlib/src/vespa/searchlib/attribute/enumstore.hpp [hunk 3]
```diff
@@ -242,6 +245,24 @@ EnumStoreT<EntryT>::compact_worst_values(bool compact_memory, bool compact_addre
...
+        _dict->compact_worst(true, false);
+        return true;
+    }
+    if (compaction_strategy.should_compact_memory(_cached_dictionary_hash_usage.usedBytes(), _cached_dictionary_hash_usage.usedBytes())) {
```
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/attribute/enumstore.hpp:259]: Another one....

File: searchlib/src/vespa/searchlib/attribute/reference_attribute.cpp [hunk 4]
```diff
@@ -304,6 +311,20 @@ ReferenceAttribute::compact_worst_values()
...
 }
 
+bool
+ReferenceAttribute::consider_compact_dictionary(const CompactionStrategy &compaction_strategy)
+{
+    auto& dictionary = _store.get_dictionary();
+    if (dictionary.has_held_buffers()) {
+        return false;
+    }
+    if (compaction_strategy.should_compact_memory(_cached_unique_store_dictionary_memory_usage.usedBytes(), _cached_unique_store_dictionary_memory_usage.usedBytes())) {
```
Comment (baldersheim) [inline searchlib/src/vespa/searchlib/attribute/reference_attribute.cpp:321]: And one more....

**Review by baldersheim — COMMENTED** [review]
Suggested fix in https://github.com/vespa-engine/vespa/pull/17436

**Pull request merged**